README.md
# vl53l1x_ros

This is STM [VL53L1X](https://www.st.com/en/imaging-and-photonics-solutions/vl53l1x.html) time-of-flight rangefinder driver for ROS. Tested on a Raspberry Pi 3 and 4 with [CJMCU-531](https://ru.aliexpress.com/item/VL53L1X/32911692450.html) board.

The code is based on the [STM VL53L1X API library](https://www.st.com/content/st_com/en/products/embedded-software/proximity-sensors-software/stsw-img007.html).

## Connecting

Example of I²C connecting VL53L1X module to Raspberry Pi:

<img src="raspberry-vl53l1x.png" width=500>

## Installation

### From package

For Raspberry Pi, there exist prebuilt Debian packages. For installation, [get the package](http://coex.space/rpi-ros-kinetic/pool/main/r/ros-kinetic-vl53l1x/) to the Raspberry and install it with `dpkg -i <package-name>.deb`.

### Manual

1. Clone the repository into your Catkin workspace:

    ```bash
    cd ~/catkin_ws/src
    git clone https://github.com/okalachev/vl53l1x_ros.git
    ```

2. Build the package:

    ```bash
    cd ~/catkin_ws
    catkin_make -DCATKIN_WHITELIST_PACKAGES="vl53l1x"
    ```

## Quick start

Run with the default settings:

```bash
rosrun vl53l1x vl53l1x_node
```

See the ranging results:

```bash
rostopic echo /vl53l1x/range
```

See the ranging rate:

```bash
rostopic hz /vl53l1x/range
```

## Parameters

All parameters are optional. Check the [`launch`](https://github.com/okalachev/vl53l1x_ros/tree/master/vl53l1x/launch) folder for launch-file templates.

* `~i2c_bus` (*int*) – I2C bus number (default: 1).
* `~i2c_address` (*int*) – I2C address (default: 0x29).
* `~mode` (*int*) – distance mode, 1 = short, 2 = medium, 3 = long (default: 3).
* `~timing_budget` (*double*) – timing budget for measurements, *s* (default: 0.1)
* `~poll_rate` (*double*) – polling data rate, *Hz* (default: 100).
* `~ignore_range_status` (*bool*) – ignore validness of measurements (default: false).
* `~min_signal` (*double*) – minimum amplitude of the signal reflected from the target to be considered valid, *MCPS* (default: 1).
* `~max_sigma` (*double*) – maximum standard deviation of the measurement to be considered valid, *m* (default: 0.015).
* `~offset` (*float*) – offset to be automatically added to measurement value, *m* (default: 0.0).
* `~frame_id` (*string*) – frame id for output `Range` messages (default: "").
* `~field_of_view` (*float*) – field of view for output `Range` messages, *rad* (default: 0.471239).
* `~min_range` (*float*) – minimum range for output `Range` messages, *m* (default: 0.0).
* `~max_range` (*float*) – maximum range for `Range` output messages, *m* (default: 4.0).

`timing_budget` is the time VL53L1X uses for ranging. The larger this time, the more accurate is measurement and the larger is maximum distance. Timing budget can be set from *0.02 s* up to *1 s*.

* *0.02 s* is the minimum timing budget and can be used only in *Short* distance mode.
* *0.033 s* is the minimum timing budget which can work for all distance modes.
* *0.14 s* is the timing budget which allows the maximum distance of *4 m* (in *Long* distance mode).

The resulting measurement rate is *1 / (timing budget + 0.004) Hz*.

Setting `ignore_range_status` to `true` makes the node to ignore `RangeStatus` field of measurements. This may significantly improve maximum distance and rate but affects quality of measurements.

`mode` is one of three distance modes, with the timing budget of *0.1 s*, *Short*, *Medium* and *Long* modes have maximum distances of 136, 290, and 360 cm, respectively.

`~min_range` and `~max_range` don't affect any device settings and only define `min_range` and `max_range` values of the output [`Range`](http://docs.ros.org/melodic/api/sensor_msgs/html/msg/Range.html) messages.

See the [official documentation](https://www.st.com/resource/en/datasheet/vl53l1x.pdf) for more information on mode and timing budget.

## Topics

### Published

* `~range` ([*sensor_msgs/Range*](http://docs.ros.org/kinetic/api/sensor_msgs/html/msg/Range.html)) – resulting measurement.
* `~data` ([*vl53l1x/MeasurementData*](https://github.com/okalachev/vl53l1x_ros/blob/master/vl53l1x/msg/MeasurementData.msg)) – additional data of the measurement.

## Copyright

Copyright © 2019 Oleg Kalachev.

Distributed under BSD 3-Clause License (https://opensource.org/licenses/BSD-3-Clause).


vl53l1x/CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package vl53l1x
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.0.0 (2021-01-28)
------------------
* Change licence to BSD-3-Clause

0.4.1 (2021-01-16)
------------------

0.4.0 (2019-12-05)
------------------
* Publish additional data of the measurements
* Publish measurement data even if range status is not valid
* Add status meanings to measurement data message file
* Implement min_signal and max_sigma parameters
* Add pass_statuses parameter for passing range statuses other than 0
* Print device info on startup
* Contributors: Alexey Rogachevskiy, Arthur Golubtsov, Oleg Kalachev


vl53l1x/src/i2c.cpp
#include <linux/i2c-dev.h>
#include <sys/ioctl.h>
#include <stdio.h>
#include <fcntl.h>
#include <arpa/inet.h>
#include <ros/ros.h>

#include "i2c.h"

int fd;

void i2c_setup(uint8_t bus, uint8_t addr)
{
	char filename[50];
	sprintf(filename, "/dev/i2c-%d", bus);

	fd = open(filename, O_RDWR);

	if (fd < 0) {
		ROS_FATAL("Failed to open I2C bus %d", bus);
		ros::shutdown();
	}

	int rc = ioctl(fd, I2C_SLAVE, addr);
	if (rc < 0) {
		ROS_FATAL("Failed to select I2C device with address 0x%02x", addr);
		ros::shutdown();
	}
}

void i2c_release()
{
	close(fd);
}

bool i2c_readRegisterMulti(uint16_t reg, size_t count, void *pdata)
{
	reg = htons(reg); // little endian to big endian

	// write register number
	if (write(fd, &reg, sizeof(reg)) < 0) {
		return false;
	}

	return read(fd, pdata, count);
}

bool i2c_readRegisterByte(int reg, uint8_t *pdata)
{
	return i2c_readRegisterMulti(reg, 1, pdata) > 0;
}

bool i2c_readRegisterWord(uint16_t reg, uint16_t *pdata)
{
	if (i2c_readRegisterMulti(reg, 2, pdata) < 0) {
		return false;
	}
	*pdata = ntohs(*pdata); // big endian to little endian
	return true;
}

bool i2c_writeRegisterMulti(uint16_t reg, size_t count, void *pdata)
{
	uint8_t buf[count + sizeof(reg)];

	reg = htons(reg); // little endian to big endian

	memcpy(buf, &reg, sizeof(reg));
	memcpy(buf + sizeof(reg), pdata, count);

	return write(fd, buf, count + sizeof(reg)) > 0;
}

bool i2c_writeRegisterByte(uint16_t reg, uint8_t value)
{
	return i2c_writeRegisterMulti(reg, 1, &value);
}


bool i2c_writeRegisterWord(uint16_t reg, uint16_t value)
{
	value = htons(value); // little endian to big endian
	return i2c_writeRegisterMulti(reg, 2, &value);
}


vl53l1x/lib/platform/vl53l1_platform.c
/* Based on: https://github.com/pimoroni/vl53l1x-python */

#include "vl53l1_platform.h"
#include "vl53l1_api.h"
#include <string.h>
#include <time.h>
#include <unistd.h>

#include "i2c.h"

VL53L1_Error VL53L1_WriteMulti(VL53L1_DEV Dev, uint16_t index, uint8_t *pdata, uint32_t count) {
	return i2c_writeRegisterMulti(index, count, pdata) ? VL53L1_ERROR_NONE : VL53L1_ERROR_CONTROL_INTERFACE;
}

// the ranging_sensor_comms.dll will take care of the page selection
VL53L1_Error VL53L1_ReadMulti(VL53L1_DEV Dev, uint16_t index, uint8_t *pdata, uint32_t count) {
	return i2c_readRegisterMulti(index, count, pdata) ? VL53L1_ERROR_NONE : VL53L1_ERROR_CONTROL_INTERFACE;
}

VL53L1_Error VL53L1_WrByte(VL53L1_DEV Dev, uint16_t index, uint8_t data) {
	return i2c_writeRegisterByte(index, data) ? VL53L1_ERROR_NONE : VL53L1_ERROR_CONTROL_INTERFACE;
}

VL53L1_Error VL53L1_WrWord(VL53L1_DEV Dev, uint16_t index, uint16_t data) {
	return i2c_writeRegisterWord(index, data) ? VL53L1_ERROR_NONE : VL53L1_ERROR_CONTROL_INTERFACE;
}

VL53L1_Error VL53L1_WrDWord(VL53L1_DEV Dev, uint16_t index, uint32_t data) {
	VL53L1_Error Status = VL53L1_ERROR_NONE;
	return Status;
}

VL53L1_Error VL53L1_UpdateByte(VL53L1_DEV Dev, uint16_t index, uint8_t AndData, uint8_t OrData) {
	VL53L1_Error Status = VL53L1_ERROR_NONE;
	return Status;
}

VL53L1_Error VL53L1_RdByte(VL53L1_DEV Dev, uint16_t index, uint8_t *data) {
	return i2c_readRegisterByte(index, data) ? VL53L1_ERROR_NONE : VL53L1_ERROR_CONTROL_INTERFACE;
}

VL53L1_Error VL53L1_RdWord(VL53L1_DEV Dev, uint16_t index, uint16_t *data) {
	return i2c_readRegisterWord(index, data) ? VL53L1_ERROR_NONE : VL53L1_ERROR_CONTROL_INTERFACE;
}

VL53L1_Error VL53L1_RdDWord(VL53L1_DEV Dev, uint16_t index, uint32_t *data) {
	VL53L1_Error Status = VL53L1_ERROR_NONE;
	return Status;
}

VL53L1_Error VL53L1_GetTickCount(
	uint32_t *ptick_count_ms)
{
	VL53L1_Error status  = VL53L1_ERROR_NONE;
	return status;
}

//#define trace_print(level, ...) \
//	_LOG_TRACE_PRINT(VL53L1_TRACE_MODULE_PLATFORM, \
//	level, VL53L1_TRACE_FUNCTION_NONE, ##__VA_ARGS__)

//#define trace_i2c(...) \
//	_LOG_TRACE_PRINT(VL53L1_TRACE_MODULE_NONE, \
//	VL53L1_TRACE_LEVEL_NONE, VL53L1_TRACE_FUNCTION_I2C, ##__VA_ARGS__)

VL53L1_Error VL53L1_GetTimerFrequency(int32_t *ptimer_freq_hz)
{
	VL53L1_Error status  = VL53L1_ERROR_NONE;
	return status;
}

VL53L1_Error VL53L1_WaitMs(VL53L1_Dev_t *pdev, int32_t wait_ms){
	usleep(wait_ms * 1000);
	return VL53L1_ERROR_NONE;
}

VL53L1_Error VL53L1_WaitUs(VL53L1_Dev_t *pdev, int32_t wait_us){
	usleep(wait_us);
	return VL53L1_ERROR_NONE;
}

VL53L1_Error VL53L1_WaitValueMaskEx(
	VL53L1_Dev_t *pdev,
	uint32_t      timeout_ms,
	uint16_t      index,
	uint8_t       value,
	uint8_t       mask,
	uint32_t      poll_delay_ms)
{
	uint8_t  register_value = 0;

	VL53L1_Error status  = VL53L1_ERROR_NONE;

	int32_t attempts = timeout_ms / poll_delay_ms;

	for(int32_t x = 0; x < attempts; x++){
		status = VL53L1_RdByte(
					pdev,
					index,
					&register_value);
		if (status == VL53L1_ERROR_NONE && (register_value & mask) == value) {
			return VL53L1_ERROR_NONE;
		}
		usleep(poll_delay_ms * 1000);
	}

	return VL53L1_ERROR_TIME_OUT;
}


vl53l1x/src/vl53l1x.cpp
/*
 * STM VL53L1X ToF rangefinder driver for ROS
 *
 * Author: Oleg Kalachev <okalachev@gmail.com>
 *
 * Distributed under BSD 3-Clause License (available at https://opensource.org/licenses/BSD-3-Clause).
 *
 * Documentation used:
 * VL53L1X datasheet - https://www.st.com/resource/en/datasheet/vl53l1x.pdf
 * VL53L1X API user manual - https://www.st.com/content/ccc/resource/technical/document/user_manual/group0/98/0d/38/38/5d/84/49/1f/DM00474730/files/DM00474730.pdf/jcr:content/translations/en.DM00474730.pdf
 *
 */

#include <string>
#include <vector>
#include <ros/ros.h>
#include <sensor_msgs/Range.h>
#include <vl53l1x/MeasurementData.h>

#include "vl53l1_api.h"
#include "i2c.h"

#define xSTR(x) #x
#define STR(x) xSTR(x)

#define CHECK_STATUS(func) { \
	VL53L1_Error status = func; \
	if (status != VL53L1_ERROR_NONE) { \
		ROS_WARN("VL53L1X: Error %d on %s", status, STR(func)); \
	} \
}

int main(int argc, char **argv)
{
	ros::init(argc, argv, "vl53l1x");
	ros::NodeHandle nh, nh_priv("~");

	sensor_msgs::Range range;
	vl53l1x::MeasurementData data;
	range.radiation_type = sensor_msgs::Range::INFRARED;
	ros::Publisher range_pub = nh_priv.advertise<sensor_msgs::Range>("range", 20);
	ros::Publisher data_pub = nh_priv.advertise<vl53l1x::MeasurementData>("data", 20);

	// Read parameters
	int mode, i2c_bus, i2c_address;
	double poll_rate, timing_budget, offset;
	bool ignore_range_status;
	std::vector<int> pass_statuses { VL53L1_RANGESTATUS_RANGE_VALID,
	                                 VL53L1_RANGESTATUS_RANGE_VALID_NO_WRAP_CHECK_FAIL,
	                                 VL53L1_RANGESTATUS_RANGE_VALID_MERGED_PULSE };

	nh_priv.param("mode", mode, 3);
	nh_priv.param("i2c_bus", i2c_bus, 1);
	nh_priv.param("i2c_address", i2c_address, 0x29);
	nh_priv.param("poll_rate", poll_rate, 100.0);
	nh_priv.param("ignore_range_status", ignore_range_status, false);
	nh_priv.param("timing_budget", timing_budget, 0.1);
	nh_priv.param("offset", offset, 0.0);
	nh_priv.param<std::string>("frame_id", range.header.frame_id, "");
	nh_priv.param("field_of_view", range.field_of_view, 0.471239f); // 27 deg, source: datasheet
	nh_priv.param("min_range", range.min_range, 0.0f);
	nh_priv.param("max_range", range.max_range, 4.0f);
	nh_priv.getParam("pass_statuses", pass_statuses);

	if (timing_budget < 0.02 || timing_budget > 1) {
		ROS_FATAL("Error: timing_budget should be within 0.02 and 1 s (%g is set)", timing_budget);
		ros::shutdown();
	}

	// The minimum inter-measurement period must be longer than the timing budget + 4 ms (*)
	double inter_measurement_period = timing_budget + 0.004;

	// Setup I2C bus
	i2c_setup(i2c_bus, i2c_address);

	// Init sensor
	VL53L1_Dev_t dev;
	VL53L1_Error dev_error;
	VL53L1_software_reset(&dev);
	VL53L1_WaitDeviceBooted(&dev);
	VL53L1_DataInit(&dev);
	VL53L1_StaticInit(&dev);
	VL53L1_SetPresetMode(&dev, VL53L1_PRESETMODE_AUTONOMOUS);

	// Print device info
	VL53L1_DeviceInfo_t device_info;
	CHECK_STATUS(VL53L1_GetDeviceInfo(&dev, &device_info));
	ROS_INFO("VL53L1X: Device name: %." STR(VL53L1_DEVINFO_STRLEN) "s", device_info.Name);
	ROS_INFO("VL53L1X: Device type: %." STR(VL53L1_DEVINFO_STRLEN) "s", device_info.Type);
	ROS_INFO("VL53L1X: Product ID: %." STR(VL53L1_DEVINFO_STRLEN) "s", device_info.ProductId);
	ROS_INFO("VL53L1X: Type: %u Version: %u.%u", device_info.ProductType,
	          device_info.ProductRevisionMajor, device_info.ProductRevisionMinor);

	// Setup sensor
	CHECK_STATUS(VL53L1_SetDistanceMode(&dev, mode));
	CHECK_STATUS(VL53L1_SetMeasurementTimingBudgetMicroSeconds(&dev, round(timing_budget * 1e6)));

	double min_signal;
	if (nh_priv.getParam("min_signal", min_signal)) {
		CHECK_STATUS(VL53L1_SetLimitCheckValue(&dev, VL53L1_CHECKENABLE_SIGNAL_RATE_FINAL_RANGE, min_signal * 65536));
	}

	double max_sigma;
	if (nh_priv.getParam("max_sigma", max_sigma)) {
		CHECK_STATUS(VL53L1_SetLimitCheckValue(&dev, VL53L1_CHECKENABLE_SIGMA_FINAL_RANGE, max_sigma * 1000 * 65536));
	}

	// Start sensor
	for (int i = 0; i < 100; i++) {
		CHECK_STATUS(VL53L1_SetInterMeasurementPeriodMilliSeconds(&dev, round(inter_measurement_period * 1e3)));
		dev_error = VL53L1_StartMeasurement(&dev);
		if (dev_error == VL53L1_ERROR_INVALID_PARAMS) {
			inter_measurement_period += 0.001; // Increase inter_measurement_period to satisfy condition (*)
		} else break;
	}

	// Check for errors after start
	if (dev_error != VL53L1_ERROR_NONE) {
		ROS_FATAL("VL53L1X: Can't start measurement: error %d", dev_error);
		ros::shutdown();
	}

	ROS_INFO("VL53L1X: ranging");

	VL53L1_RangingMeasurementData_t measurement_data;

	// Main loop
	ros::Rate r(poll_rate);
	while (ros::ok()) {
		r.sleep();
		range.header.stamp = ros::Time::now();

		// Check the data is ready
		uint8_t data_ready = 0;
		VL53L1_GetMeasurementDataReady(&dev, &data_ready);
		if (!data_ready) {
			continue;
		}

		// Read measurement
		VL53L1_GetRangingMeasurementData(&dev, &measurement_data);
		VL53L1_ClearInterruptAndStartMeasurement(&dev);

		// Publish measurement data
		data.header.stamp = range.header.stamp;
		data.signal = measurement_data.SignalRateRtnMegaCps / 65536.0;
		data.ambient = measurement_data.AmbientRateRtnMegaCps / 65536.0;
		data.effective_spad = measurement_data.EffectiveSpadRtnCount / 256;
		data.sigma = measurement_data.SigmaMilliMeter / 65536.0 / 1000.0;
		data.status = measurement_data.RangeStatus;
		data_pub.publish(data);

		// Check measurement for validness
		if (!ignore_range_status &&
		    std::find(pass_statuses.begin(), pass_statuses.end(), measurement_data.RangeStatus) == pass_statuses.end()) {
			char range_status[VL53L1_MAX_STRING_LENGTH];
			VL53L1_get_range_status_string(measurement_data.RangeStatus, range_status);
			ROS_DEBUG("Range measurement status is not valid: %s", range_status);
			ros::spinOnce();
			continue;
		}

		// Publish measurement
		range.range = measurement_data.RangeMilliMeter / 1000.0 + offset;
		range_pub.publish(range);

		ros::spinOnce();
	}

	// Release
	ROS_INFO("VL53L1X: stop ranging");
	VL53L1_StopMeasurement(&dev);
	i2c_release();
}
